pageserver/src/tenant/timeline.rs

```diff
@@ -5700,6 +5709,9 @@ impl<'a> TimelineWriter<'a> {
         if state.prev_lsn == Some(lsn) {
             // Rolling mid LSN is not supported by downstream code.
             // Hence, only roll at LSN boundaries.
+            //
+            // TODO: could we have more hints here what actually goes wrong? is it just that we
+            // assume the next layer starts from where the previous layer ends?
```

Comment:
The layer map keeps track of the LSN at which to open the next layer (this is due to the fact that requests for a new layer are serviced by the layer manager who owns the layer map). Each time a layer is frozen `LayerMap::next_open_layer_at` is updated to `lsn_at_which_we_have_frozen + 1`. Hence, if we tried to roll mid lsn, the new in memory layer would have an off by one start lsn.

The invariant for the in memory layer is that it does not contain any entries with lsns below `InMemoryLayer::start_lsn`. When the invariant doesn't hold it can trigger the assertion in `InMemoryLayer::freeze` and reads below `InMemoryLayer::start_lsn` would also fail.


This is why it doesn't work code-wise, but there's also the issue of name clashes. Since L0s use the full key range, the lsn range encoded in their name has to be unique to avoid overwrites. I think this is the core of the issue and why the code is how it is.